### ISSUE TITLE
GORM cascades validation for associations more aggressively than before

### ISSUE THREAD

**Issue by longwa** [grails/grails-data-mapping#1106]
In Grails 2.5.x (whatever GORM version that used), the HibernateDomainClassValidator had a short circuit to skip cascading validation like so:
```
if (associatedDomainClass == null || !isOwningInstance(bean, associatedDomainClass) && !persistentProperty.isExplicitSaveUpdateCascade()) {
    return;
}
```

For an association like:
```
class Author {
    Publisher publisher
}
class Publisher {
    String name
}
```
The publisher association would not be validated unless it were actually dirty (even if the proxy were unwrapped).

In Grails 3.3.5 (GORM 6.1.9), the default behavior will now validate the Publisher if it's just loaded (unwrapped) by the Author, even if it hasn't changed at all. This is because the check for `isExplicitSaveUpdateCascade` has been replaced with a check for:

`!association.doesCascade(CascadeType.PERSIST, CascadeType.MERGE)`

However, ALL to-one associations, by default, are `CascadeType.PERSIST`, which ends up continuing past the new check:

```
if (associatedEntity == null || (!association.isOwningSide() && !association.doesCascade(CascadeType.PERSIST, CascadeType.MERGE) )) {
    return
}
```

This causes way more validations in our application for data which really doesn't need to be validated at all since it's not even dirty and is really just reference data or other non-owned data.

Non-owned objects shouldn't cascade their validations if they aren't dirty themselves. If they are dirty, they will be validated anyway in the `ClosureEventListener.doValidate` callback when they are being saved.

I think the check should be changed to just be:
```
if (associatedEntity == null || (!association.isOwningSide() && !association.doesCascade(CascadeType.ALL))) {
    return
}
```

That way, if explicit ALL cascade is given, it will cascade the validation, otherwise it will only validate owning relationships and things that are actually dirty and being saved by cascade.

### Environment Information
GORM 6.1.9

### Example Application
https://github.com/longwa/grails-issue-cascadeValidateG3

**Comment by zoladkow** [comment]
`cascadeValidate` is not described anywhere in the docs.

**Comment by zoladkow** [comment]
@longwa You had a good idea, but realization is efective only on the root entity of the graph. Or the original entity on which `.save()` was called. 

Now, imagine we have 
```groovy
class Filter {
  static hasMany = [conditions: Condition]
  
  static mapping = {
    conditions column: 'FILTR_ID', joinTable: false
  }
}

// a filter has multiple conditions, related to specific attribute
class Condition {
  static belongsTo = Filter
  
  Attribute attribute
  
  static mapping = {
    attribute column: 'ATTR_ID', cascadeValidate: false
  }
}

class Attribute {
  ...
}
```
With this model, we have a preexisting Attribute entity, which we refer too from a Condition created as part of a Filter entity. We then do `newFilter.save()`
Now, what will happen when validation arrives to 'attribute' property? Turns out the method `doesCascadeValidate("attribute")` will look for this property on the Filter entity not the current validated entity. And this is logically incorrect!

It does not even matter if I persist the `Condtion` entity before assigning to `Filter` since validation done during `filter.save()` will still cascade through the graph blissfully ignoring the `cascadeValidate` one level deep.

Basically what I'm saying is that `doesCascadeValidate()` should also consider `associatedEntity` mappings or even better - `doesCascadeValidate()` should be implemented on `Association`

**Comment by longwa** [comment]
I’ve actually reworked this some to allow different options. I need to submit a new PR with documentation.
